### ISSUE TITLE
RFC: change `dynlib` handling on linux to use standard dynamic linking instead of runtime loading via `dlopen` 

### ISSUE THREAD

**Issue by arnetheduck** [nim-lang/RFCs#58]
In the current scheme, when using `dynlib` on linux, Nim will generate code that loads the library via `dlopen` and successive `dlsym` calls. This is problematic for several reasons:

* linking on linux should typically be done using the `soname` - linking via filename appears to work on developer system but will cause issues when binary is used on a machine where the named file is symlinked to a different version: see also http://tldp.org/HOWTO/Program-Library-HOWTO/shared-libraries.html - note that typically, when passing a library name to C compiler via `-lXXX`, the compiler will translate the `XXX` (linker convenience name) to the `soname` as part of the linking process - in nim, one has to be extra careful because wrapping libraries without using header files may lead to subtle ABI mismatches.

* opening library via `dlopen` means a larger binary that takes longer to load - the generated code is verbose and slow compared to what `ld` can do using the special linking sections in the executable

* linking via `dlopen` means tools like `ldd`, `gdb` and others show incomplete information about application dependencies, making dependency analysis harder than it has to be

* using plain `libXXX.so` file (without versioning information, as given by soname) requires the plain `.so` to be installed both when running and when compiling the program - this plain file is a symlink that's only installed by `-devel` packages that users typically don't keep installed - this makes it harder to both distribute and compile nim programs. it also promotes situations where things work on developer computers but not on end-user systems, leading to unnecessary friction in the release process.

Instead, I'd like to suggest that Nim uses standard linux linking when using `dynlib` - that will solve all of the above problems: fewer ABI issues, faster startup/load times, smaller binaries, standard tools work and no devel packages are needed.

**Comment by FedericoCeratto** [comment]
Echoing what @arnetheduck wrote:
Scripted languages like Python and Ruby most likely use dlopen because they are not compiled.
For a compiled language, using dlopen unnecessarily and making the dependencies invisible to ldd is unexpected.
Also, ignoring/bypassing the header files at compile times can be risky.

**Comment by jasagiri** [comment]
Title: Proposal: A Theoretical Framework for "Vendor Libraries" (solving the 3-layer
  FFI problem)

  I have been following the discussion on RFC #58 and believe it's missing a crucial
  component: a standardized way to handle vendored (bundled) libraries.

  In my technical analysis, "Solving Nim's C FFI Problems"
  (https://zenn.dev/jasagiri/articles/c41657b012ccac), I identified that FFI issues
  occur across 3 distinct layers, and we must solve all of them to make "Standard
  Dynamic Linking" (the goal of this RFC) viable for real-world applications.

  The 3-Layer Framework
   1. Layer 1 (The Linker): Recording dependencies (e.g., DT_NEEDED).
   2. Layer 2 (The Loader/dlopen): Resolving the library filename (prefixes like lib and
      extensions like .so).
   3. Layer 3 (The Search Path): Telling the OS where to find the file (RPATH, $ORIGIN,
      @executable_path).

  Why RFC #58 is incomplete without this:
  RFC #58 focuses on Layer 1 (moving from dlopen to standard linking). However, if we
  change how Nim links without a standard way to handle Layer 2 and 3 for bundled
  libraries, we break the portability that Nim users love.

  I have developed a Proof-of-Concept in nim-vendor-lib
  (https://github.com/jasagiri/nim-vendor-lib) that automates these layers.

  What this looks like in practice:

  The Problem (Current/Before):
  Users manually write brittle, platform-specific code to guess paths and extensions,
  often failing to handle dependencies of the library itself.

   1 const path = when defined(posix): "./vendor/lib/libfoo.so" else: "foo.dll"
   2 proc foo() {.importc, dynlib: path.}
   3 # Result: Works for foo, but fails if foo needs libbar.so in the same folder.

  The Solution (Proposed/After):
  A standardized "Vendor" logic that resolves the path and automatically injects the
  correct RPATH at compile-time.

   1 import std/vendor_lib
   2 # 1. Resolves "foo" to libfoo.so/dylib/dll (Layer 2)
   3 # 2. Injects RPATH $ORIGIN/vendor/lib (Layer 3)
   4 proc foo() {.importc, dynlib: vendorDynlib("foo").}

  Proposal for Nim Core:
  To fulfill the goal of RFC #58, I propose integrating this 3-layer awareness into Nim:
   1. Standardize a "vendor" directory (e.g., vendor/lib).
   2. Automate RPATH injection based on that directory during compilation.
   3. Provide a compiler-level vendor: scheme for {.dynlib.} that handles the
      platform-agnostic path resolution.

  By solving the "Vendor" problem formally, we make the transition to standard dynamic
  linking safe and professional for everyone.

**Comment by Araq** [comment]
Not gonna happen. You basically propose that we change the UNIX way of using libs in /usr/lib to something else.

**Comment by jasagiri** [comment]
I fully agree with the goal of RFC #58 to prioritize standard dynamic linking over runtime dlopen for better system integration.
 
Regarding the concern that nim-vendor-lib "deviates from UNIX standards," I would like to clarify its intent. This tool leverages the official ELF standard feature $ORIGIN (RPATH) to resolve libraries relative to the executable. This is not a deviation, but rather an utilization of standard OS loader mechanisms to support relocatable, self-contained application bundles.
  
While the FHS (standard system paths) is ideal for core system binaries, vendoring libraries is a common and necessary practice for third-party application distribution (e.g., AppImage, proprietary software, or portable CLI tools) to avoid ABI mismatches across different distributions.
  
I am using this approach for my own projects and would appreciate it if this could be considered as a reference solution for developers who need to ship standalone binaries without forcing system-wide dependencies.

**Comment by Araq** [comment]
"better system integration" -- well it's only one broken tool that forgot about the fact that dynamic libraries can be loaded -- well -- dynamically, `strace` has no such problem and again Python uses dlopen and became one of the most used PLs regardless...

As for "ship standalone binaries", that's fundamentally not supported on Unix via this "we'll distribute apps over /usr, /etc" madness which is why every OS has multiple appstores these days that said "we are not gonna do this shit" and rightly so.

**Comment by jasagiri** [comment]
Thank you for the insightful feedback. I truly appreciate the perspective on the traditional UNIX philosophy, where shared libraries are managed at the system level by package managers.

Through this discussion, I've realized that we are looking at two different layers of the software lifecycle:

1. The Linking Layer (Scope of RFC #58): How Nim integrates with the OS and adheres to system standards (FHS). This is crucial for system-level tools and official distribution packages.
2. The Distribution Layer (Scope of nim-vendor-lib): How developers ship portable, standalone binaries to users across diverse environments with varying library versions.

Regarding the point about Python, it is interesting to note that while Python itself uses dlopen, the industry standard for distributing binary wheels on Linux (the manylinux standard) relies heavily on tools like auditwheel. These tools automatically bundle dependencies and inject RPATH/$ORIGIN into the binaries to ensure they work regardless of the host's system libraries. This is exactly the "distribution layer" problem I aimed to solve.

I now agree that this functionality belongs in the application-distribution/tooling space rather than being a core part of the Nim language or its standard linking behavior.

I will continue to maintain nim-vendor-lib as a specialized utility for developers who need to ship self-contained binaries, while fully supporting the move towards standard dynamic linking for Nim's core. Thank you for helping me clarify these distinct requirements.

To provide more technical context regarding the Python example:

In the Python ecosystem, the industry standard for distributing binary packages (Wheels) on Linux is defined by the "manylinux" specifications. A core component of this standard is a tool called auditwheel.

When a developer prepares a binary for distribution, auditwheel performs a "repair" process that:
  1. Identifies external shared library dependencies.
  2. Copies those libraries into a dedicated internal folder within the package.
  3. Uses patchelf to inject RPATH entries using $ORIGIN to ensure the loader finds these bundled libraries first.

This demonstrates that even for Python, dlopen alone is insufficient for reliable distribution across different Linux environments. The Python community explicitly relies on the same "Vendoring + RPATH" strategy to bypass the constraints of the traditional UNIX filesystem hierarchy (FHS) and ensure portability.

My goal with nim-vendor-lib is simply to provide Nim developers with a similar, standard-compliant mechanism for the distribution layer, without interfering with the system-level linking standards addressed in RFC #58.

Additionally, regarding the mention of strace:

I see strace as an essential diagnostic tool—an "examination kit" used to identify why a library failed to load after a failure occurs. However, the purpose of nim-vendor-lib is to act as a "preventative" measure.

By properly configuring the distribution layer with RPATH and bundled libraries, the application is designed to "just work" out of the box. This spares the end-user from ever needing to use tools like strace or ldd to resolve environment-specific path issues.

Importantly, this approach does not hide anything from the OS; it remains fully transparent to strace. It simply ensures that the OS loader finds the intended dependencies automatically, following the same successful, portable distribution patterns used by major ecosystems like Python.
